src/corelibs/U2View/src/ov_msa/MSAEditor.h

```diff
@@ -48,7 +48,7 @@ class PairwiseAlignmentWidgetsSettings {
     QString algorithmName;
     bool inNewWindow;
     QString resultFileName;
-    PairwiseAlignmentTask *pairwiseAlignmentTask;
```

Comment:
I see that 'pairwiseAlignmentTask' sometimes is deleted by 3rd party code and you want to be protected using QPointer and this is OK.
Question: which code deleted the task?

Reply:
https://github.com/ugeneunipro/ugene/blob/83018b0ad6624087bf8b1251e19b6ba3b52a8a79/src/corelibs/U2Private/src/TaskSchedulerImpl.cpp#L927

Reply:
This is true for all tasks in UGENE: all tasks are deleted by the task scheduler. 
But any code that holds a reference to a task usually is required to listen for the task state changes like run/report/finished, etc. and update the pointer to the task: null it when task is finished.

I see that this kind of the code is written here: it listens to state changes:
```connect(task, SIGNAL(si_stateChanged()), SLOT(sl_alignComplete()));) but does it incorrectly.```
It nulls the pointer when task is finished (before it is deleted).

This way your fix may hide the original problem and leave some invalid code unfixed just because it does not crash anymore. I'm OK with QPointer (as I said) just want to understand the original problem.

Question: what would be the fix with no QPointer. Why we are keeping a pointer to a deleted task and do not null it in in sl_alignComplete?

Reply:
Because if we close the "Pairwise Alignment" tab, the object of the `PairAlign` class is deleted, which should have null the current `pairwiseAlignmentTask` in `pairwiseAlignmentWidgetsSettings`. These settings are a shared object (via a simple pointer) between `PairAlign` and `MSAEditor`: https://github.com/ugeneunipro/ugene/blob/0addaa0f72c9f484244b3ddf1c258bc38d1e278a/src/corelibs/U2View/src/ov_msa/PairAlign/PairAlign.cpp#L71 When we open the "Pairwise Alignment" tab again, we create a new `PairAlign` object, the `pairwiseAlignmentWidgetsSettings` of which are taken from `MSAEditor` -- the outdated settings with a nonnull pointer: https://github.com/ugeneunipro/ugene/blob/0addaa0f72c9f484244b3ddf1c258bc38d1e278a/src/corelibs/U2Gui/src/options_panel/OptionsPanel.cpp#L136

Could there be any other problems here, do you think?

Reply:
This sounds correct. Thank you for the explanation.